src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -174,9 +174,29 @@ public void StartMonitoringNetworkChanges()
                 return;
             }
 
-            using (ExecutionContext.SuppressFlow())
+            // RFC: https://tools.ietf.org/html/rfc7838#section-2.2
+            //    When alternative services are used to send a client to the most
+            //    optimal server, a change in network configuration can result in
+            //    cached values becoming suboptimal.  Therefore, clients SHOULD remove
+            //    from cache all alternative services that lack the "persist" flag with
+            //    the value "1" when they detect such a change, when information about
+            //    network state is available.
+            try
+            {
+                using (ExecutionContext.SuppressFlow())
+                {
+                    NetworkChange.NetworkAddressChanged += networkChangedDelegate;
+                }
+            }
+            catch (Exception e)
             {
-                NetworkChange.NetworkAddressChanged += networkChangedDelegate;
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, $"Exception when subscribing to NetworkChange.NetworkAddressChanged: {e}");
+
+                // We can't monitor network changes, so technically "information
+                // about network state is not available" and we can just keep
+                // all Alt-Svc entries until their expiration time.
+                //
+                // keep the _networkChangeCleanup field null so we don't try again needlessly
```

Comment:
Can the cleanup be unhappy if we fail to add it?

Reply:
Yeah, I think we might need to prevent both disposal and finalization of _networkChangeCleanup...

Reply:
I checked only the Unix code, but the unregister call should be safe (it does not try to allocate a new socket, nor does it rely on the handler being previously registered in the first place)

Reply:
I checked other platform implementations and all are robust against unregistering callbacks which was not registered before.